src/contactsmodel.cpp

```diff
@@ -0,0 +1,187 @@
+    this->contacts.clear();
+    QListIterator<QString> userIdIterator(contactIds);
+    while (userIdIterator.hasNext()) {
+        QString nextUserId = userIdIterator.next();
+        LOG("Hydrating contact:" << nextUserId);
+        this->contacts.append(this->tdLibWrapper->getUserInformation(nextUserId));
+    }
+    LOG("Hydrated contacts:" << this->contacts.size());
+    std::sort(this->contacts.begin(), this->contacts.end(), compareUsers);
+}
+
+void ContactsModel::applyFilter(const QString &filter)
+{
+    LOG("Applying filter:" << filter);
+    beginResetModel();
+    this->filter = filter;
+    this->filteredContacts.clear();
+    if (!this->filter.isEmpty()) {
+        QListIterator<QVariant> contactIterator(this->contacts);
+        while (contactIterator.hasNext()) {
+            QVariantMap contact = contactIterator.next().toMap();
+            if (contact.value(LAST_NAME).toString().contains(this->filter, Qt::CaseInsensitive)) {
+                this->filteredContacts.append(contact);
+                continue;
+            }
+            if (contact.value(FIRST_NAME).toString().contains(this->filter, Qt::CaseInsensitive)) {
+                this->filteredContacts.append(contact);
+                continue;
+            }
+            if (contact.value(USERNAME).toString().contains(this->filter, Qt::CaseInsensitive)) {
+                this->filteredContacts.append(contact);
+                continue;
+            }
+        }
+    }
+    endResetModel();
+}
+
+void ContactsModel::synchronizeContacts()
+{
+    LOG("Synchronizing device contacts");
+    QVariantList deviceContacts;
+    QSqlQuery databaseQuery(this->deviceContactsDatabase);
+    databaseQuery.prepare("select distinct c.contactId, c.firstName, c.lastName, n.phoneNumber from Contacts as c inner join PhoneNumbers as n on c.contactId = n.contactId where n.phoneNumber is not null and ( c.firstName is not null or c.lastName is not null );");
+    if (databaseQuery.exec()) {
+        LOG("Device contacts successfully selected from database!");
+        while (databaseQuery.next()) {
+            QVariantMap singleContact;
+            singleContact.insert("first_name", databaseQuery.value(1).toString());
+            singleContact.insert("last_name", databaseQuery.value(2).toString());
+            singleContact.insert("phone_number", databaseQuery.value(3).toString());
+            deviceContacts.append(singleContact);
+            LOG("Found contact" << singleContact.value("first_name").toString() << singleContact.value("last_name").toString() << singleContact.value("phone_number").toString());
+        }
+        if (!deviceContacts.isEmpty()) {
+            LOG("Importing found contacts" << deviceContacts.size());
+            this->tdLibWrapper->importContacts(deviceContacts);
+        }
+        emit contactsSynchronized();
+    } else {
+        LOG("Error selecting contacts from database!");
+        emit errorSynchronizingContacts();
```

Comment:
A strange way of returning a boolean value from a function by emitting a signal. Or did you mean that this function will later become asynchronous?

Reply:
It was originally intended to be asynchronous, but as it was fast enough, I kept it synchronous and forgot to adjust those thingies here. Will be changed...